tests/base64_test.c

```diff
@@ -47,39 +49,68 @@ int base64_test(void)
 #ifdef LTC_BASE64_URL
    const struct {
       const char* s;
-      int is_strict;
+      int flag;
    } url_cases[] = {
-         {"vuiSPKIl8PiR5O-rC4z9_xTQKZ0", 0},
-         {"vuiSPKIl8PiR5O-rC4z9_xTQKZ0=", 1},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0==", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0===", 0},
-         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0====", 0},
-         {"vuiS*=PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
-         {"vuiS*==PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
-         {"vuiS*===PKIl8P*iR5O-rC4*z9_xTQKZ0=", 0},
+         {"vuiSPKIl8PiR5O-rC4z9_xTQKZ0", strict},                       /* 0 */
+         {"vuiSPKIl8PiR5O-rC4z9_xTQKZ0=", strict},
+         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0", insane},
+         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0=", insane},
+         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0==", insane},
+         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0===", insane},                 /* 5 */
+         {"vuiS*PKIl8P*iR5O-rC4*z9_xTQKZ0====", insane},
+         {"vuiS*=PKIl8P*iR5O-rC4*z9_xTQKZ0=", insane},
+         {"vuiS*==PKIl8P*iR5O-rC4*z9_xTQKZ0=", insane},
+         {"vuiS*==\xffPKIl8P*iR5O-rC4*z9_xTQKZ0=", insane},
+         {"vuiS PKIl8P\niR5O-rC4\tz9_xTQKZ0", relaxed},                 /* 10 */
+         {"vuiS PKIl8P\niR5O-rC4\tz9_xTQKZ0=", relaxed},
+         {"vuiS PKIl8P\niR5O-rC4\tz9_xTQKZ0==", relaxed},
+         {"vuiS PKIl8P\niR5O-rC4\tz9_xTQKZ0===", relaxed},
+         {"vuiS PKIl8P\niR5O-rC4\tz9_xTQKZ0====", relaxed},
+         {"vuiS\rPKIl8P\niR5O-rC4\tz9_xTQKZ0=", relaxed},               /* 15 */
+         {"vuiS\rPKIl8P\niR5O-rC4\tz9_xTQKZ0= = =\x00", relaxed},
+         {"\nvuiS\rPKIl8P\niR5O-rC4\tz9_xTQKZ0=\n", relaxed},
+         {"vuiSPKIl8PiR5O-rC4z9_xTQK", invalid},
    };
 
    for (x = 0; x < sizeof(url_cases)/sizeof(url_cases[0]); ++x) {
        slen1 = strlen(url_cases[x].s);
        l1 = sizeof(tmp);
-       if(url_cases[x].is_strict)
+       if(url_cases[x].flag == strict) {
           DO(base64url_strict_decode(url_cases[x].s, slen1, tmp, &l1));
-       else
+          DO(do_compare_testvector(tmp, l1, special_case, sizeof(special_case) - 1, "base64url_strict_decode", x));
+          DO(base64url_sane_decode(url_cases[x].s, slen1, tmp, &l1));
+          DO(do_compare_testvector(tmp, l1, special_case, sizeof(special_case) - 1, "base64url_sane_decode/strict", x));
+          DO(base64url_decode(url_cases[x].s, slen1, tmp, &l1));
+          DO(do_compare_testvector(tmp, l1, special_case, sizeof(special_case) - 1, "base64url_decode/strict", x));
+       }
+       else if(url_cases[x].flag == relaxed) {
+          DO(base64url_strict_decode(url_cases[x].s, slen1, tmp, &l1) == CRYPT_INVALID_PACKET ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR);
```

Comment:
I already thought about it while writing the last base16 testcase... should we add this?

```
diff --git a/tests/common.h b/tests/common.h
index 1ba8820..f63e26b 100644
--- a/tests/common.h
+++ b/tests/common.h
@@ -18,2 +18,3 @@ extern prng_state yarrow_prng;
 #define DOX(x, str) do { fprintf(stderr, "%s - %s:\n", #x, (str)); run_cmd((x), __LINE__, __FILE__, #x, (str)); } while (0)
+#define SHOULD_FAIL(x) do { fprintf(stderr, "%s:\n", #x); run_cmd((x) != CRYPT_OK ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR, __LINE__, __FILE__, #x, NULL); } while (0)
 #else
@@ -21,2 +22,3 @@ extern prng_state yarrow_prng;
 #define DOX(x, str) do { run_cmd((x), __LINE__, __FILE__, #x, (str)); } while (0)
+#define SHOULD_FAIL(x) do { run_cmd((x) != CRYPT_OK ? CRYPT_OK : CRYPT_FAIL_TESTVECTOR, __LINE__, __FILE__, #x, NULL); } while (0)
 #endif
```

Reply:
Maybe, but perhaps as `SHOULD_FAIL(x, expected_retval)`

Reply:
I also thought about `EXPECT(x, e)` but in my opinion that's another use-case.
Most cases where we do negative tests shouldn't care about the error-code returned, it should just be ensured that the result is not OK.